{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d906121c",
   "metadata": {},
   "source": [
    "# 1. Data Loading and Preprocessing\n",
    "\n",
    "In this section, we will load the data from provided URLs and perform some basic preprocessing steps. The datasets include gene expression data, mutation data, copy number variation data, and subtype covariates.\n",
    "\n",
    "##  üêº1.1Data Loading\n",
    "\n",
    "üêº`pandas` Pandas is a powerful Python library that is used for data loading and preprocessing. \n",
    "\n",
    "It helps with reading datasets, handling missing values, and combining data.\n",
    "\n",
    "We will load the following datasets from [https://raw.githubusercontent.com/WanbingZeng/OMINEX/main/data/ ]\n",
    "1. **Gene Expression Data (COREAD_gex.csv)**\n",
    "2. **Mutation Data (COREAD_mu.csv)**\n",
    "3. **Copy Number Variation Data (COREAD_cn.csv)**\n",
    "4. **Subtype Covariates (COREAD_subtypes.csv)**\n",
    "## üêº1.1.1 Gene expression\n",
    "\n",
    "##### Example Codeüìã \n",
    "\n",
    "```python\n",
    "import pandas as pd\n",
    "# Base URL for data download\n",
    "base_url = \"https://raw.githubusercontent.com/WanbingZeng/OMINEX/main/data/\"\n",
    "# Load Gene Expression Data\n",
    "x1 = pd.read_csv(base_url + \"COREAD_gex.csv\")\n",
    "# Display the DataFrame\n",
    "x1\n",
    "\n",
    "```\n",
    "##### ‚ñ∂Ô∏è Please run the code below and see how `x1` looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b9048bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aec89660",
   "metadata": {},
   "source": [
    "‚ö†Ô∏è**We can see that the first column appears as `Unnamed: 0`, which indicates that simply using `pd.read_csv` does not fix the first row as the index. Therefore, we need set the first row as the index using  `index_col=0`.**\n",
    "\n",
    "‚ö†Ô∏è**After loading the data, it is important to check the format of the data, including the row and column names.**\n",
    "#####  Now let us reload `COREAD_gex.csv` with `index_col=0`\n",
    "##### Example Codeüìã \n",
    "```python\n",
    "\n",
    "x1 = pd.read_csv(base_url + \"COREAD_gex.csv\",index_col=0)\n",
    "x1\n",
    "```\n",
    "##### ‚ñ∂Ô∏è Reload COREAD_gex.csv as `x1` \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d85d27",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "00705181",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "The dataset provided contains gene expression levels across multiple cases. Each row represents a different gene, and each column corresponds to a different case or sample. The values in the matrix represent the expression level of each gene in each case. The values represent the expression levels of each gene in different cases. Higher values indicate higher gene expression.\n",
    "\n",
    "##### Data Structure\n",
    "\n",
    "- **Rows**: Each row represents a different **gene**\n",
    " `RNF113A`\n",
    " `S100A13`\n",
    " `AP3D1`\n",
    "\n",
    "- **Columns**: Each column represents a different **patient**\n",
    "   `Case1`\n",
    "  `Case2`\n",
    "  `Case3`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e08cab8",
   "metadata": {},
   "source": [
    "## üêº1.1.2 Mutaion \n",
    "##### ‚ñ∂Ô∏è Load **COREAD_mu.csv** as `x2`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a0f8c6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "64bb4836",
   "metadata": {},
   "source": [
    "#### Mutation Data Description\n",
    "\n",
    "In the mutation matrix, each row represents a different gene, and each column represents a different case or sample. The values in the matrix indicate the presence (`1`) or absence (`0`) of a mutation for each gene in each case.\n",
    "##### Data Structure\n",
    "\n",
    "- **Rows**: Each row corresponds to a **gene**:\n",
    "   `TTN`\n",
    "   `TP53`\n",
    "   `APC`\n",
    "\n",
    "\n",
    "- **Columns**: Each column represents a different **patient**\n",
    "  `Case1`  `Case2`  `Case3`\n",
    "\n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd2e62ff",
   "metadata": {},
   "source": [
    "## üêº1.1.3 Copy Number Variation\n",
    "##### ‚ñ∂Ô∏è Load **COREAD_cn.csv** as `x3`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c143a1a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "881ea91a",
   "metadata": {},
   "source": [
    "In the CNV matrix, each row represents a different copy number variation (CNV) region, and each column represents a different case or sample. The values in the matrix indicate the presence (`1`) or absence (`0`) of a CNV for each region in each case.\n",
    "\n",
    "##### Data Structure\n",
    "\n",
    "- **Rows**: Each row corresponds to a **CNV region**:\n",
    "  `8p23.2`\n",
    "  `8p23.3`\n",
    "  `8p23.1`\n",
    "- **Columns**: Each column represents a different **patient**:\n",
    "  `Case1`\n",
    "  `Case2`\n",
    "  `Case3`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba5c5645",
   "metadata": {},
   "source": [
    "## üêº1.1.4 Cancer subtypes\n",
    "These tumors belong to two molecular subtypes, CMS1 and CMS3, as defined by the Colorectal Cancer Subtyping Consortium.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa7b3eb4",
   "metadata": {},
   "source": [
    "##### Example Codeüìã \n",
    "```python\n",
    "covariates = pd.read_csv(base_url + \"COREAD_subtypes.csv\")\n",
    "covariates\n",
    "```\n",
    "##### ‚ñ∂Ô∏è Please run the code below and see how `Subtype Covariates` looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f11a44c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8473c8bf",
   "metadata": {},
   "source": [
    "‚ö†Ô∏è**We can see that the first column appears as `Unnamed: 0` , which indicates that simply using `pd.read_csv` does not fix the first row as the index. Therefore, we need set the first row as the index using `index_col=0` **\n",
    "\n",
    "‚ö†Ô∏è**We can see that the first row appears as `0` , which indicates that simply using `pd.read_csv` does not fix the first column as the index. Therefore, we need set the first column as the index using `header=0` **\n",
    "#####  Now let us reload `COREAD_covariates.csv` with `index_col=0` and `header=0`\n",
    "##### Example Codeüìã \n",
    "```python\n",
    "covariates = pd.read_csv(base_url + \"COREAD_subtypes.csv\",index_col=0, header=0)\n",
    "covariates\n",
    "```\n",
    "##### ‚ñ∂Ô∏è Reload COREAD_gex.csv as `x1` \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1eb91a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "757f9d8e",
   "metadata": {},
   "source": [
    "The covariates matrix contains clinical and molecular information for each patient. Each row represents a different case, with columns providing details on demographics, clinical stage, mutation status, and molecular subtypes.\n",
    "\n",
    "##### ü©∫Data Structure\n",
    "\n",
    "- **Rows**: Each row corresponds to a **patient**:\n",
    "   `Case1`\n",
    "   `Case2`\n",
    "   `Case3`\n",
    "\n",
    "- **Columns**: Each column represents a different type of information of the patients, in this project, we only use the first column and the  <mark>**subtypes**</mark> column:\n",
    "  - `dataset`: Data source ( `tcga: the cancer genome atlas`)\n",
    "  - `age`: Patient's age\n",
    "  - `gender`: Patient's gender\n",
    "  - `stage`: Clinical stage of the disease\n",
    "  - `pt`: Tumor size\n",
    "  - `pn`: Regional lymph node involvement\n",
    "  - `pm`: Distant metastasis\n",
    "  - `tnm`: TNM classification\n",
    "  - `grade`: Tumor grade\n",
    "  - `msi`: Microsatellite instability status\n",
    "  - `cimp`: CpG island methylator phenotype\n",
    "  - `kras_mut`: KRAS mutation status\n",
    "  - `braf_mut`: BRAF mutation status\n",
    "  - <mark>**subtypes**</mark>: Molecular subtype (e.g., `CMS1`, `CMS3`)\n",
    "  - `osMo`: Overall survival months\n",
    "  - `osStat`: Overall survival status\n",
    "  - `rfsMo`: Recurrence-free survival months\n",
    "  - `rfsStat`: Recurrence-free survival status\n",
    "\n",
    "##### ü©∫CMS1 and CMS3\n",
    "CMS1 and CMS3 are two of the four consensus molecular subtypes of colorectal cancer (CRC) identified by the Consensus Molecular Subtyping \n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f0641d",
   "metadata": {},
   "source": [
    "##### For ü©∫`covariates`, since we are only using the `subtypes` column in this project, we need to only keep the `subtypes` column using:\n",
    "  **`data[['column']]`**: Selects  the only `column`  from the `data` DataFrame.\n",
    "##### Example Codeüìã \n",
    "\n",
    "\n",
    "\n",
    "```python\n",
    "covariates_s = covariates[['subtypes']]\n",
    "covariates_s\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "##### ‚ñ∂Ô∏è remove  columns except for `subtypes` from the `covariates` DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d2211e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "69c17c1f",
   "metadata": {},
   "source": [
    "Congratulations! Your have loaded all your datasets!üéâ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac92551",
   "metadata": {},
   "source": [
    "## 1.2 Data preprocessing\n",
    "From 1.1 Data Loading, we have read the datatset as\n",
    "1. `x1`**Gene Expression Data (COREAD_gex.csv)**\n",
    "2. `x2`**Mutation Data (COREAD_mu.csv)**\n",
    "3. `x3`**Copy Number Variation Data (COREAD_cn.csv)**\n",
    "4. `covariates`**Subtype Covariates (COREAD_subtypes.csv)**\n",
    "\n",
    "Now we need to check if there is missing values and merge the data for the analyzing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e3f1dec",
   "metadata": {},
   "source": [
    "## üêº1.2.1 Missing Values\n",
    "\n",
    "\n",
    "We will still use the üêº`pandas` library. Here is an example to check for missing values in `x1`, which shows the number of missing values in each column. \n",
    "\n",
    "##### Example Codeüìã \n",
    "\n",
    "```python\n",
    "print(x1.isna().sum())  \n",
    "\n",
    "```\n",
    "##### ‚ñ∂Ô∏è Check if `x1`, `x2`, `x3` , `covariates_s` have missing vaues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dfee58a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "aadff01c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "943ab807",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c68e284",
   "metadata": {},
   "outputs": [],
   "source": [
    "#covariates_s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0433506e",
   "metadata": {},
   "source": [
    "If there is no missing values, let's go on."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7ca3536",
   "metadata": {},
   "source": [
    "## üêº1.2.2 Merge the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04bf330f",
   "metadata": {},
   "source": [
    "##### Example üé¨ \n",
    "Here shows two datasets with fixed row and column names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "1c51717c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case1</th>\n",
       "      <th>Case2</th>\n",
       "      <th>Case3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GeneA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneB</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Case1  Case2  Case3\n",
       "GeneA      1      0      1\n",
       "GeneB      0      1      1\n",
       "GeneC      1      1      0"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1 = {\n",
    "    'Case1': [1, 0, 1],\n",
    "    'Case2': [0, 1, 1],\n",
    "    'Case3': [1, 1, 0]\n",
    "}\n",
    "d1 = pd.DataFrame(data1, index=['GeneA', 'GeneB', 'GeneC'])\n",
    "d1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "9a17b891",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Case1</th>\n",
       "      <td>Y1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case2</th>\n",
       "      <td>Y2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case3</th>\n",
       "      <td>Y1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Category\n",
       "Case1       Y1\n",
       "Case2       Y2\n",
       "Case3       Y1"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the second DataFrame (covariates)\n",
    "data_covariates = {\n",
    "    'Category': ['Y1','Y2', 'Y1'],\n",
    "}\n",
    "c = pd.DataFrame(data_covariates, index=['Case1', 'Case2', 'Case3'])\n",
    "c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8874d2ff",
   "metadata": {},
   "source": [
    "##### üîóStep1 Transpose d1 : `.T`\n",
    "‚ö†Ô∏èNormally, we set features(e.g `genes`) as row names and samples(e.g `patiens`) as column names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "153cb415",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "      <th>GeneC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Case1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case3</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneA  GeneB  GeneC\n",
       "Case1      1      0      1\n",
       "Case2      0      1      1\n",
       "Case3      1      1      0"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transpose x1 to match the column names of c\n",
    "d1_T=d1.T\n",
    "d1_T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e730d34",
   "metadata": {},
   "source": [
    "##### üîóStep2 Merge dataframes on the `case` column \n",
    "\n",
    "`pd.merge()`:  Combines two DataFrames based on a common key. Here, instead of using columns as keys, we use the indexes.\n",
    "\n",
    "`left_index=True`: Indicates that the index of the left DataFrame (x1_T) should be used for merging.\n",
    "\n",
    "`right_index=True`: Indicates that the index of the right DataFrame (c) should be used for merging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5a895037",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GeneA  GeneB  GeneC Category\n",
      "Case1      1      0      1       Y1\n",
      "Case2      0      1      1       Y2\n",
      "Case3      1      1      0       Y1\n"
     ]
    }
   ],
   "source": [
    "merged_data = pd.merge(d1_T, c, left_index=True,right_index=True)\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb1dd33d",
   "metadata": {},
   "source": [
    "##### Please use `.T` and `pd.merge` to combine the datasets below, we will use the combined data to select the important features.\n",
    "1. `x1`**Gene Expression Data (COREAD_gex.csv)**\n",
    "2. `x2`**Mutation Data (COREAD_mu.csv)**\n",
    "3. `x3`**Copy Number Variation Data (COREAD_cn.csv)**\n",
    "4. `covariates`**Subtype Covariates (COREAD_subtypes.csv)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21f9727c",
   "metadata": {},
   "source": [
    "##### ‚ñ∂Ô∏è Now, try to combine the dataframes: `x1` and `covariates_s`  as `m1` ,we will use the combined data to select the features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e0ab2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0ddc1bb3",
   "metadata": {},
   "source": [
    "##### ‚ñ∂Ô∏è  combine the dataframes: `x2` and `covariates_s`  as `m2` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df25f007",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b4d5ea88",
   "metadata": {},
   "source": [
    "##### ‚ñ∂Ô∏è  combine the dataframes: `x3` and `covariates_s`  as `m3` ,we will use the combined data to select the features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b04e5b79",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aadc4d0b",
   "metadata": {},
   "source": [
    "###### ‚ùì How to combine all the datasets \n",
    "  ANSWER : We can use `pd.contact` to combine `x1`, `x2`, `x3` first. Then run the same process.\n",
    "##### Example Codeüìã \n",
    "```python\n",
    "combined_data = pd.concat([x1,x2,x3], axis=0)\n",
    "combined_data_T=combined_data.T\n",
    "combined_all=pd.merge(combined_data_T, covariates_s,left_index=True, right_index= True)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eee69b7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd00fb9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5480d60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "713e89e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55712040",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2359ebc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c20c8a2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "734d5ce7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4c5bf4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d54068a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c21714f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fb106b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "88b256a9",
   "metadata": {},
   "source": [
    "üîë ...recommended to go through the entire process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "69486b2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNF113A</th>\n",
       "      <th>S100A13</th>\n",
       "      <th>AP3D1</th>\n",
       "      <th>ATP6V1G1</th>\n",
       "      <th>UBQLN4</th>\n",
       "      <th>TPPP3</th>\n",
       "      <th>TSSC4</th>\n",
       "      <th>FOS</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>CHRAC1</th>\n",
       "      <th>...</th>\n",
       "      <th>15q13.2</th>\n",
       "      <th>7q21.13</th>\n",
       "      <th>13q31.1</th>\n",
       "      <th>4q35.2</th>\n",
       "      <th>7q32.1</th>\n",
       "      <th>17p11.2</th>\n",
       "      <th>15q11.1</th>\n",
       "      <th>15q11.2</th>\n",
       "      <th>18q12.1</th>\n",
       "      <th>subtypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Case1</th>\n",
       "      <td>21.195670</td>\n",
       "      <td>19.726005</td>\n",
       "      <td>11.530217</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.356366</td>\n",
       "      <td>12.767472</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16.489747</td>\n",
       "      <td>17.708855</td>\n",
       "      <td>16.056793</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case2</th>\n",
       "      <td>21.508658</td>\n",
       "      <td>18.657292</td>\n",
       "      <td>12.988300</td>\n",
       "      <td>14.126750</td>\n",
       "      <td>19.622082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21.274977</td>\n",
       "      <td>18.115900</td>\n",
       "      <td>12.230063</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case3</th>\n",
       "      <td>20.080716</td>\n",
       "      <td>18.970339</td>\n",
       "      <td>10.837587</td>\n",
       "      <td>15.313250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.374816</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.785306</td>\n",
       "      <td>11.779289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.883356</td>\n",
       "      <td>10.242480</td>\n",
       "      <td>19.792995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.031538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.077533</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.077866</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case117</th>\n",
       "      <td>11.673601</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.527430</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16.734721</td>\n",
       "      <td>11.906196</td>\n",
       "      <td>11.835825</td>\n",
       "      <td>10.059176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case118</th>\n",
       "      <td>11.717598</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.580360</td>\n",
       "      <td>11.986711</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.445359</td>\n",
       "      <td>11.950195</td>\n",
       "      <td>11.879824</td>\n",
       "      <td>11.102489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case119</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.146131</td>\n",
       "      <td>14.633627</td>\n",
       "      <td>18.139138</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.063835</td>\n",
       "      <td>18.105979</td>\n",
       "      <td>12.294651</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.768502</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case120</th>\n",
       "      <td>13.003646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.870922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.360792</td>\n",
       "      <td>12.651393</td>\n",
       "      <td>12.581015</td>\n",
       "      <td>10.803979</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case121</th>\n",
       "      <td>11.363485</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18.007650</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.043740</td>\n",
       "      <td>17.254570</td>\n",
       "      <td>12.595831</td>\n",
       "      <td>12.525453</td>\n",
       "      <td>12.747811</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CMS1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows √ó 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           RNF113A    S100A13      AP3D1   ATP6V1G1     UBQLN4      TPPP3  \\\n",
       "Case1    21.195670  19.726005  11.530217   0.000000  15.356366  12.767472   \n",
       "Case2    21.508658  18.657292  12.988300  14.126750  19.622082   0.000000   \n",
       "Case3    20.080716  18.970339  10.837587  15.313250   0.000000   0.000000   \n",
       "Case4     0.000000  11.883356  10.242480  19.792995   0.000000   0.000000   \n",
       "Case5     0.000000  12.077533   0.000000   0.000000   0.000000   0.000000   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "Case117  11.673601   0.000000   0.000000   0.000000  13.527430   0.000000   \n",
       "Case118  11.717598   0.000000   0.000000  20.580360  11.986711   0.000000   \n",
       "Case119   0.000000  13.146131  14.633627  18.139138   0.000000  14.063835   \n",
       "Case120  13.003646   0.000000  19.870922   0.000000   0.000000   0.000000   \n",
       "Case121  11.363485   0.000000  18.007650   0.000000   0.000000  11.043740   \n",
       "\n",
       "             TSSC4        FOS      ERBB3     CHRAC1  ...  15q13.2  7q21.13  \\\n",
       "Case1     0.000000  16.489747  17.708855  16.056793  ...      0.0      0.0   \n",
       "Case2     0.000000  21.274977  18.115900  12.230063  ...      0.0      0.0   \n",
       "Case3    22.374816   0.000000  19.785306  11.779289  ...      0.0      0.0   \n",
       "Case4     0.000000  13.031538   0.000000   0.000000  ...      0.0      0.0   \n",
       "Case5    11.077866   0.000000   0.000000   0.000000  ...      0.0      0.0   \n",
       "...            ...        ...        ...        ...  ...      ...      ...   \n",
       "Case117  16.734721  11.906196  11.835825  10.059176  ...      0.0      1.0   \n",
       "Case118  14.445359  11.950195  11.879824  11.102489  ...      0.0      1.0   \n",
       "Case119  18.105979  12.294651   0.000000  12.768502  ...      1.0      0.0   \n",
       "Case120  15.360792  12.651393  12.581015  10.803979  ...      0.0      0.0   \n",
       "Case121  17.254570  12.595831  12.525453  12.747811  ...      0.0      0.0   \n",
       "\n",
       "         13q31.1  4q35.2  7q32.1  17p11.2  15q11.1  15q11.2  18q12.1  subtypes  \n",
       "Case1        1.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS1  \n",
       "Case2        0.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS1  \n",
       "Case3        1.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS1  \n",
       "Case4        0.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS1  \n",
       "Case5        0.0     0.0     0.0      0.0      0.0      0.0      1.0      CMS1  \n",
       "...          ...     ...     ...      ...      ...      ...      ...       ...  \n",
       "Case117      1.0     0.0     1.0      1.0      0.0      0.0      1.0      CMS3  \n",
       "Case118      1.0     0.0     1.0      1.0      0.0      0.0      1.0      CMS3  \n",
       "Case119      0.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS3  \n",
       "Case120      0.0     0.0     0.0      0.0      0.0      0.0      0.0      CMS1  \n",
       "Case121      0.0     1.0     0.0      1.0      0.0      0.0      0.0      CMS1  \n",
       "\n",
       "[121 rows x 801 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "base_url = \"https://raw.githubusercontent.com/WanbingZeng/OMINEX/main/data/\"\n",
    "# Data Loading\n",
    "x1 = pd.read_csv(base_url + \"COREAD_gex.csv\",index_col=0)\n",
    "x2 = pd.read_csv(base_url + \"COREAD_mu.csv\",index_col=0)\n",
    "x3 = pd.read_csv(base_url + \"COREAD_cn.csv\",index_col=0)\n",
    "covariates = pd.read_csv(base_url + \"COREAD_subtypes.csv\",index_col=0,header=0)\n",
    "covariates_s = covariates[['subtypes']]\n",
    "# Data processing\n",
    "x1_T=x1.T\n",
    "x2_T=x2.T\n",
    "x3_T=x3.T\n",
    "m1=pd.merge(x1_T, covariates_s,left_index=True, right_index= True)\n",
    "m2=pd.merge(x2_T, covariates_s,left_index=True, right_index= True)\n",
    "m3=pd.merge(x3_T, covariates_s,left_index=True, right_index= True)\n",
    "combined_data = pd.concat([x1,x2,x3], axis=0)\n",
    "combined_data_T=combined_data.T\n",
    "combined_all=pd.merge(combined_data_T, covariates_s,left_index=True, right_index= True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
